index out of range when availableIn=0 and r.needOutput is true

Trace:

```
r.bufferRead = 131072, r.availableIn = 0, inputPosition = 131072, buffer len = 131072
panic: runtime error: index out of range

goroutine 140 [running]:
github.com/itchio/butler/vendor/gopkg.in/kothar/brotli-go.v0/dec.(*BrotliReader).Read(0xc42016e150, 0xc42011d640, 0x1, 0x1, 0x1, 0x0, 0x0)
        /home/amos/go/src/github.com/itchio/butler/vendor/gopkg.in/kothar/brotli-go.v0/dec/decode.go:143 +0x62d
github.com/itchio/butler/vendor/github.com/itchio/wharf/wire.(*ReadContext).ReadByte(0xc42001b980, 0x10080, 0x0, 0x0)
        /home/amos/go/src/github.com/itchio/butler/vendor/github.com/itchio/wharf/wire/read.go:33 +0x52
encoding/binary.ReadUvarint(0xf915c0, 0xc42001b980, 0x4138ed, 0x9ac080, 0xc420010300)
        /usr/local/go/src/encoding/binary/varint.go:110 +0x4a
github.com/itchio/butler/vendor/github.com/itchio/wharf/wire.(*ReadContext).ReadMessage(0xc42001b980, 0xf97bc0, 0xc4201961c0, 0x0, 0x0)
        /home/amos/go/src/github.com/itchio/butler/vendor/github.com/itchio/wharf/wire/read.go:63 +0x3b
github.com/itchio/butler/vendor/github.com/itchio/wharf/pwr.readOps(0xc42001b980, 0xc420010300, 0xc4200102a0)
        /home/amos/go/src/github.com/itchio/butler/vendor/github.com/itchio/wharf/pwr/apply.go:936 +0x114
created by github.com/itchio/butler/vendor/github.com/itchio/wharf/pwr.(*ApplyContext).patchAll
        /home/amos/go/src/github.com/itchio/butler/vendor/github.com/itchio/wharf/pwr/apply.go:466 +0x6ff
amos@poppy:~/butler-tests/possessed$

```

Offending bit of code:

```go
                if r.availableIn > 0 || r.needOutput {
                        // Decompress
                        inputPosition := r.bufferRead - int(r.availableIn)
                        fmt.Printf("r.bufferRead = %d, r.availableIn = %d, inputPosition = %d, buffer len = %d\n", r.bufferRead, r.availableIn, inputPosition, len(r.buffer));
                        result := C.BrotliDecompressStream_Wrapper(
                                &r.availableIn,
                                (*C.uint8_t)(unsafe.Pointer(&r.buffer[inputPosition])),
                                &availableOut,
                                (*C.uint8_t)(unsafe.Pointer(&p[0])),
                                &r.totalOut,
                                (*C.BrotliState)(r.state),
                        )

                        n = maxOutput - int(availableOut)
```

I think the issue is that.. `&r.buffer[inputPosition]` is valid in C because there's no bound check going on, but in golang there is, and it fails. I wonder if `unsafe.Pointer(&r.buffer)+inputPosition` would work (and be faster!

If that fails, maybe we can pass a null pointer instead in that very case.

I'm impressed this even happened though (I have a reproducible test case here, but it involves a >2GB file right now), I've been using brotli-go in production for 1yr+ without issues :)

--- a/dec/decode.go
+++ b/dec/decode.go
@@ -136,9 +136,13 @@ func (r *BrotliReader) Read(p []byte) (n int, err error) {
 		if r.availableIn > 0 || r.needOutput {
 			// Decompress
 			inputPosition := r.bufferRead - int(r.availableIn)
+			nextIn := unsafe.Pointer(uintptr(0))
+			if r.availableIn > 0 {
+				nextIn = unsafe.Pointer(&r.buffer[inputPosition])
+			}
 			result := C.BrotliDecompressStream_Wrapper(
 				&r.availableIn,
-				(*C.uint8_t)(unsafe.Pointer(&r.buffer[inputPosition])),
+				(*C.uint8_t)(nextIn),
 				&availableOut,
 				(*C.uint8_t)(unsafe.Pointer(&p[0])),
 				&r.totalOut,

--- a/dec/decode.go
+++ b/dec/decode.go
@@ -136,7 +136,7 @@ func (r *BrotliReader) Read(p []byte) (n int, err error) {
 		if r.availableIn > 0 || r.needOutput {
 			// Decompress
 			inputPosition := r.bufferRead - int(r.availableIn)
-			nextIn := unsafe.Pointer(uintptr(0))
+			nextIn := unsafe.Pointer(nil)
 			if r.availableIn > 0 {
 				nextIn = unsafe.Pointer(&r.buffer[inputPosition])
 			}

